File: Lab_12/src/TextEditor.java
import java.util.ArrayList;
import java.util.List;

public class TextEditor {
    private String content;
    private int cursorPosition;
    private List<String> selections;

    public TextEditor() {
        this.content = "";
        this.cursorPosition = 0;
        this.selections = new ArrayList<>();
    }

    public void setContent(String content) {
        this.content = content;
    }

    public void setCursorPosition(int position) {
        if (position >= 0 && position <= content.length()) {
            this.cursorPosition = position;
        }
    }

    public void addSelection(String selection) {
        selections.add(selection);
    }

    public void clearSelections() {
        selections.clear();
    }

    public EditorMemento createMemento() {
        return new EditorMemento(content, cursorPosition, new ArrayList<>(selections));
    }

    public void restoreFromMemento(EditorMemento memento) {
        this.content = memento.getContent();
        this.cursorPosition = memento.getCursorPosition();
        this.selections = new ArrayList<>(memento.getSelections());
    }

    public String getContent() {
        return content;
    }
}

File: Lab 12 Memento/Lab-12_Memento/TextEditor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab_12_Memento
{
    public class TextEditor
    {
        private string content;
        private int cursorPosition;
        private List<string> selections;


        public TextEditor()
        {
            content = string.Empty;
            cursorPosition = 0;
            selections = new List<string>();
        }

        public void SetContent(string newContent)
        {
            content = newContent;
            cursorPosition = content.Length;
        }

        public void SetCursorPosition(int position)
        {
            cursorPosition = position;
        }

        public void AddSelection(string selection)
        {
            selections.Add(selection);
        }


        public void ClearSelection(string selection)
        {
            selections.Clear();
        }

        public EditorMemento CreateMemento()
        {
            return new EditorMemento(content, cursorPosition, selections);
        }   


        public void RestoreFromMemento(EditorMemento memento)
        {
            content = memento.Content;
            cursorPosition = memento.CursorPosition;
            selections = new List<string>(memento.Selections);
        }


        public void DisplayState()
        {
            Console.WriteLine("Content: " + content);
            Console.WriteLine("Cursor Position: " + cursorPosition);
            Console.WriteLine("");

        }
    }
}
